pygerduty/v2.py

```diff
@@ -0,0 +1,582 @@
+import copy
+import re
+
+import six
+from six.moves import urllib
+from common import (
+    Requester,
+    _lower,
+    _upper,
+    _singularize,
+    _pluralize,
+    _json_dumper,
+)
+
+__author__ = "Gary M. Josack <gary@dropbox.com>"
+from .version import __version__, version_info  # noqa
+
+TRIGGER_LOG_ENTRY_RE = re.compile(
+    r'log_entries/(?P<log_entry_id>[A-Z0-9]+)'
+)
+
+# TODO:
+# Support for Log Entries
+# Support for Reports
+
+
+class Error(Exception):
+    pass
+
+
+class BadRequest(Error):
+    def __init__(self, payload, *args, **kwargs):
+        # Error Reponses don't always contain all fields.
+        # Sane defaults must be set.
+        self.code = payload.get("error", {}).get('code', 99999)
+        self.errors = payload.get("error", {}).get('errors', [])
+        self.message = payload.get("error", {}).get('message', str(payload))
+
+        Error.__init__(self, *args, **kwargs)
+
+    def __str__(self):
+        return "{0} ({1}): {2}".format(
+            self.message, self.code, self.errors)
+
+
+class NotFound(Error):
+    pass
+
+
+class Collection(object):
+    paginated = True
+
+    def __init__(self, pagerduty, base_container=None):
+        self.name = getattr(self, "name", False) or _lower(self.__class__.__name__)
+        self.sname = getattr(self, "sname", False) or _singularize(self.name)
+        self.container = (getattr(self, "container", False) or
+                          globals()[_upper(self.sname)])
+
+        self.pagerduty = pagerduty
+        self.base_container = base_container
+
+    def create(self, **kwargs):
+        path = "{0}".format(self.name)
+        if self.base_container:
+            path = "{0}/{1}/{2}".format(
+                self.base_container.collection.name,
+                self.base_container.id, self.name)
+
+        data = {self.sname: {}}
+
+        # requester_id needs to be up a level
+        if "requester_id" in kwargs:
+            extra_headers = {"From": kwargs["requester_id"]}
+            del kwargs["requester_id"]
+
+        data[self.sname] = kwargs
+
+        response = self.pagerduty.request("POST", path, data=_json_dumper(data), extra_headers=extra_headers)
```

Comment:
extra_headers only conditionally exists so this could throw and attribute error. You probably want to do something like

```
        hdrs = {} 
        if "requester_id" in kwargs:
            hdrs["From"] = : kwargs.pop("requester_id")
 
```

So that we always have a variable defined. This needs fixing in other locations as well.